### ISSUE TITLE
Introduce APIs to represent a resolved dependency graph in a lazy manner

### ISSUE THREAD

**Issue by eskatos** [gradle/gradle#12871]
`FileCollection` is lazy and allows to reason about a flat set of files, possibly coming from dependency resolution.

`ArtifactCollection` is lazy and allows to reason about a flat set of artifacts, coming from dependency resolution.

But there's no types that are lazy and allows to reason about the dependency graph.

For instance `ResolutionResult` contains the information about the result of dependency resolution, `ResolvedComponentResult` represents a component instance in the resolved dependency graph, but they aren't lazy.

This issue is about introducing APIs that allow reasoning about the dependency graph in a lazy manner. This would allow to capture dependency graphs in a lazy way for usage in task inputs, worker action parameters or build service parameters.

It could be new types, or `Provider<?>` alike.

### Context

With configuration caching tasks can only hold lazy types. In other words tasks that need to reason about the dependency graph are not supported without such APIs.

Tasks such as `:dependencyInsight` or `GenerateGraphTask` in `ResolveTestFixtures` are examples of tasks that currently can't be implemented supporting instant execution.

**Comment by melix** [comment]
Can you clarify what you mean by not lazy? `Configuration#getIncoming` returns a `ResolvableDependencies` instance which is indeed lazy.

**Comment by eskatos** [comment]
Thanks @melix!

Right, `ResolvableDependencies` is lazy and is the entry point to get views over the dependency graph as `FileCollection`, `ArtifactCollection` or `ResolutionResult`.

My understanding is that you don't want to use `ResolvableDependencies` as, say, a task input because it represents too much, but instead use one of the three "views" depending on your use case. `ResolutionResult` is the one "view" that isn't lazy.

One requirement we have with instant execution is that we need to be able to (de)serialize the "spec" of the task inputs. And `ResolvableDependencies` is somehow tangled with `Configuration` instances which makes it hard to capture.

**Comment by melix** [comment]
Ok I think I kind of understand what you mean. Currently there's no way to say "the _graph_ is the input" of a task. It's either a file collection or an artifact collection. I had a use case where I actually didn't care about the _files_ but only about the graph and currently it's not possible to represent that as a task input. As for `ResolutionResult` not being lazy... that again I don't clearly understand, because how could it be lazy if it _is_ the resolution result? Note that even if you have a `ResolutionResult`, the actual _files_ are resolved lazily. This area of the code is particularly complex, because of laziness... and the fact there are actually multiple APIs available (old result API, new result API) and no unified API to get both the resolution result _and_ artifacts.

**Comment by eskatos** [comment]
> currently it's not possible to represent that [the graph] as a task input

Yes, that's exactly what I'm after.

Currently you can use a `Configuration` as a task input, and then get to the graph in your task action. That's what tasks like `:dependencyInsight` or `GenerateTaskGraph` are doing. But with instant execution `Configuration` isn't supported as a task input because it is way too much heavyweight. 

> As for ResolutionResult not being lazy... that again I don't clearly understand, because how could it be lazy if it is the resolution result?

It is lazy wrt files, but not wrt dependency resolution. Correct ?

I guess what we'd need would be something like a `Provider<ResolutionResult>` that we could understand enough so we can capture its spec in order to defer the dependency resolution.

> no unified API to get both the resolution result and artifacts

I assume that this is another different use case.

**Comment by autonomousapps** [comment]
I don't know enough about gradle core to comment intelligibly on the "what is lazy" discussion, but I _do_ have a complex gradle plugin that uses `Configuration` objects as task inputs, and I understand from Paul that these are incompatible with instant execution. I'm going to point to places where I use them in case these concrete use-cases help resolve the issue.

In [AbiAnalysisAggregateReportTask](https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/blob/master/src/main/kotlin/com/autonomousapps/tasks/AbiAnalysisAggregateReportTask.kt#L24), the configuration in question is a custom one created by the plugin to produce artifacts by subprojects for consumption by the root project. The code looks like:
```kotlin
  @get:PathSensitive(PathSensitivity.RELATIVE)
  @get:InputFiles
  lateinit var abiReports: Configuration

  @TaskAction fun action() {
    val abiAnalysisReports = abiReports.dependencies.map { dependency ->
      val path = (dependency as ProjectDependency).dependencyProject.path

      val abiList = abiReports.fileCollection(dependency).singleFile.readLines()

      path to abiList
    }.toMap()
```

[AdviceAggregateReportTask](https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/blob/master/src/main/kotlin/com/autonomousapps/tasks/AdviceAggregateReportTask.kt#L27) and [DependencyMisuseAggregateReportTask](https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/blob/master/src/main/kotlin/com/autonomousapps/tasks/DependencyMisuseAggregateReportTask.kt#L26) do something very similar (a theme is developing).

Perhaps more interestingly, [DependencyMisuseTask](https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/blob/master/src/main/kotlin/com/autonomousapps/tasks/DependencyMisuseTask.kt#L39) uses a "normal" configuration to get access to the graph. It uses it like this:
```kotlin
@TaskAction fun action() {
  val resolvedComponentResult: ResolvedComponentResult = runtimeConfiguration
    .incoming
    .resolutionResult
    .root

  // ...time passes...
  // `Dependency` is a a kotlin data class, not from gradle core

  private fun Dependency.asResolvedDependencyResult(): ResolvedDependencyResult? =
    root.dependencies.filterIsInstance<ResolvedDependencyResult>().find { rdr ->
      identifier == rdr.selected.id.asString()
    }

  // and that is used here, to connect unused _declared_ dependencies to used _transitive_ dependencies
  // Connect used-transitives to direct dependencies
  val unusedDepsWithTransitives: Set<UnusedDirectComponent> =
    unusedDeps.mapNotNullToSet { unusedDep ->
      unusedDep.asResolvedDependencyResult()?.let { rdr ->
        relate(
          unusedDependency = rdr,
          unusedDirectComponent = UnusedDirectComponent(unusedDep, mutableSetOf()),
          usedTransitiveComponents = usedTransitiveComponents
        )
      }
    }
}
```
and finally, [DependencyReportTask](https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/blob/master/src/main/kotlin/com/autonomousapps/tasks/DependencyReportTask.kt#L46) does this with a `Configuration`:
```kotlin
/**
 * Traverses the top level of the dependency graph to get all "direct" dependencies.
 */
private fun Configuration.directDependencies(): Set<Dependency> {
  // Update all-artifacts list: transitive or not?
  // runtime classpath will give me only the direct dependencies
  val dependencies: Set<DependencyResult> =
    incoming
      .resolutionResult
      .root
      .dependencies

  return traverseDependencies(dependencies)
}

/**
 * This was heavily modified from code found in the Gradle 5.6.x documentation. Can't find the link any more.
 */
private fun traverseDependencies(results: Set<DependencyResult>): Set<Dependency> = results
  .filterIsInstance<ResolvedDependencyResult>()
  .mapToSet { result ->
    val componentResult = result.selected

    when (val componentIdentifier = componentResult.id) {
      is ProjectComponentIdentifier -> Dependency(componentIdentifier)
      is ModuleComponentIdentifier -> Dependency(componentIdentifier)
      else -> throw GradleException("Unexpected ComponentIdentifier type: ${componentIdentifier.javaClass.simpleName}")
    }
  }
```

**Comment by autonomousapps** [comment]
How does configuration caching work with `@Internal` inputs? I have a task with the following:
```kotlin
  /**
   * This is the "official" input for wiring task dependencies correctly, but is otherwise
   * unused.
   */
  @get:Classpath
  lateinit var artifactFiles: FileCollection

  /**
   * This is what the task actually uses as its input.
   */
  @get:Internal
  lateinit var runtimeConfiguration: Configuration
```
and later, in the task action:
```kotlin
val resolvedComponentResult: ResolvedComponentResult = runtimeConfiguration
  .incoming
  .resolutionResult
  .root
```

**Comment by eskatos** [comment]
@autonomousapps, configuration cache treats all task properties the same, they all are task state. `@Internal` is for up-to-date checks / build cache. If you try your code above with the configuration cache it'll report that `var runtimeConfiguration: Configuration` is not supported.

**Comment by stephanenicolas** [comment]
Are you sure @eskatos that we can use `ArtifactCollection` already has an input for a task ? It seems to me it's not working, only file collection seems to work.

**Comment by eygraber** [comment]
Is there anything planned for this? [It's blocking using the configuration cache on Android.](https://issuetracker.google.com/issues/162074215)

**Comment by eskatos** [comment]
This will be available in Gradle 7.5.

The new API can be tested using a 7.5 nightly and this repository that provides examples: https://github.com/adammurdoch/dependency-graph-as-task-inputs
